tests/pytorch/test_numerics.py

```diff
@@ -1575,3 +1575,137 @@ def test_kv_cache_accuracy(dtype, bs, model_key, use_RoPE, input_format, module,
+        out = torch.split(torch.randn(m, n, dtype=dtype, device="cuda"), m_splits) # output
+        grad = False
+    elif layout == "NN":
+        A = [torch.randn(n, k, dtype=dtype, device="cuda") for _ in range(z)]      # weight
+        B = torch.split(torch.randn(m, n, dtype=dtype, device="cuda"), m_splits)   # grad_output
+        out = torch.split(torch.randn(m, k, dtype=dtype, device="cuda"), m_splits) # dgrad
+        grad = True
+    else:  # layout == "NT"
+        A = torch.split(torch.randn(m, k, dtype=dtype, device="cuda"), m_splits) # input
+        B = torch.split(torch.randn(m, n, dtype=dtype, device="cuda"), m_splits) # grad_output
+        out = [torch.randn(n, k, dtype=dtype, device="cuda") for _ in range(z)]  # wgrad
+        grad = True
+
+    out_ref = _torch_grouped_gemm(A, B, out, layout, accumulate)
+    grouped_gemm(
+        A,
+        B,
+        out,
+        dtype,
+        get_multi_stream_cublas_workspace(),
+        grad=grad,
+        accumulate=accumulate,
+        layout=layout,
+    )
+    for o, o_ref in zip(out, out_ref):
+        torch.testing.assert_close(o, o_ref)
+
+
+@pytest.mark.parametrize("shape", [(1, 128, 128, 512),
+                                   (8, 1024, 128, 512),
+                                   (16, 4096, 128, 512),])
+@pytest.mark.parametrize("fp8_dtype", [tex.DType.kFloat8E4M3, tex.DType.kFloat8E5M2])
+@pytest.mark.parametrize("accumulate", [False, True])
+def test_fp8_grouped_gemm(shape, fp8_dtype, accumulate):
+    if not fp8_available:
+        pytest.skip(reason_for_no_fp8)
+
+    z, m, k, n = shape
+    m_splits = m // z
+
+    dtype = torch.bfloat16
+    A = [torch.randn(n, k, dtype=dtype, device="cuda") for _ in range(z)]      # weight
+    B = torch.split(torch.randn(m, k, dtype=dtype, device="cuda"), m_splits)   # input
+    out = torch.split(torch.randn(m, n, dtype=dtype, device="cuda"), m_splits) # output
+    out_ref = [o.clone() for o in out]
+
+    # fp8 should be robust enough to this fake scale
+    scale = 1 + torch.rand(z * 3, dtype=torch.float32, device="cuda")
+    scale_inv = 1 / scale
+    amax = torch.zeros(1024, z * 3, dtype=torch.float32, device="cuda")
+
+    A_fp8 = [torch.ops.tex_ts.cast_to_fp8_ts(
+        A[i],
+        scale,
+        amax,
+        scale_inv,
+        i,  # fp8 meta tensor index
+        tex.DType.kFloat8E4M3,
+    ) for i in range(z)]
+    B_fp8 = [torch.ops.tex_ts.cast_to_fp8_ts(
+        B[i],
+        scale,
+        amax,
+        scale_inv,
+        z + i,  # fp8 meta tensor index
+        fp8_dtype,
+    ) for i in range(z)]
+
+    fp8_grouped_gemm(
+        A_fp8,
+        scale_inv,
+        0,  # A_offset
+        tex.DType.kFloat8E4M3,
+        B_fp8,
+        scale_inv,
+        z,  # B_offset
+        fp8_dtype,
+        out,
+        dtype,
+        get_multi_stream_cublas_workspace(),
+        accumulate=accumulate,
+    )
+
+    # baseline
+    for i in range(z):
+        fp8_gemm(
```

Comment:
Hi, 
Would it be better to use the normal `gemm` as a baseline instead of the `fp8_gemm`?

Reply:
Using `fp8_gemm` here can ensure bit-wise match, which is stricter compared to normal `gemm` I think. I updated the tols to demonstrate that.
I also updated bf16/fp16/fp32 grouped gemm tests to use TE's `gemm` instead of torch's to ensure bit-wise match.

Reply:
I also updated how normal `gemm` handle A/B matrices don't have any elements. Previously, when A/B matrices don't have any elements, it directly returns the D matrix, which could be created with `torch.empty()` and contains garbage values in it.